rxjava/src/main/kotlin/com/mindera/skeletoid/rxjava/Observable.kt

```diff
@@ -38,29 +45,50 @@ fun <T : Any> Observable<T>.createUniqueConcurrentRequestCache(requestMap: Concu
 /**
  * Holds an item and an exception. Used in [delayAtLeast]
  */
-internal data class DataHolder<T>(val something: T? = null, val throwable: Throwable? = null)
+internal data class DataHolder<T>(val something: T?= null, val throwable: Throwable? = null)
 
 /**
  * Extension that always waits for at least timeToWait in the provided timeUnit before emitting. This will happen even if an exception is thrown. Used to prevent loadings to just blink when a connection is too fast.
  *
  * @param timeToWait - the time to wait
  * @param timeUnit - the unit of timeToWait
  */
-fun <T> Observable<T>.delayAtLeast(timeToWait: Long = 1000, timeUnit: TimeUnit = TimeUnit.MILLISECONDS): Observable<T> {
+fun <T> Observable<T>.delayAtLeast(
+    timeToWait: Long = 1000,
+    timeUnit: TimeUnit = TimeUnit.MILLISECONDS
+): Observable<T> {
+    var data : DataHolder<T>
     return Observable.zip<DataHolder<T>, Long, DataHolder<T>>(this.map {
-        DataHolder(
-            something = it
-        )
-    }.onErrorReturn { DataHolder(throwable = it) },
-            Observable.timer(timeToWait, timeUnit), BiFunction { t, _ -> t }).map { it.something ?: it.throwable?.let { throwable -> throw throwable } ?: throw Exception() }
+        data = DataHolder(something = it)
+        data
+    }
+        .onErrorReturn { DataHolder(throwable = it) },
+        Observable.timer(timeToWait, timeUnit), BiFunction { t, _ -> t })
+        .map {
+            it.something ?: throw it.throwable ?: Exception()
+        }
 }
 
 fun <T> Observable<T>.allowMultipleSubscribers(): Observable<T> =
-        share()
-                .replay(1)
-                .autoConnect(1)
-
+    share()
+        .replay(1)
+        .autoConnect(1)
+
+fun <T> Observable<T>.filterAndDo(condition: Boolean, elseFunction: () -> Unit): Observable<T> =
+    doOnNext {
+        if (condition) {
+            elseFunction()
+        }
+    }
+        .filter { condition }
```

Comment:
can't we do:
this.filter { condition }.doOnNext { elseFunction() }
?

Reply:
The parameter name made you think it's a different thing. It not to be run as else, but when the condition is met.
So it should be before or it won't run. I've changed the name to `functionToRunIfConditionIsMet`

Reply:
@neteinstein I didn't even noticed the parameter name (really else function is the exact opposite of what it is doing).
When you do .filter { condition }, it will only "let through" the inputs when the condition is true, so it will should be exactly as if there was an if(condition) and only continue the chain if the condition is true. The resulting observable would only then be composed of the inputs where the condition is true. 
So doing onNext in that Observable is the same as onNext with an if condition in the unfiltered observable, without checking the condition twice.

Reply:
Test 1:
        Observable.just(1,2,3,4,5,6,7,8,9,10).filter { it % 2 == 0 }.doOnNext { Log.v(TAG, "Test 2 Number $it is pair") }.subscribe {
            Log.v(TAG, "Test 2 Subscribe $it")
        }.addTo(disposables)
Output:
Test Number 2 is pair
Test Subscribe 2
Test Number 4 is pair
Test Subscribe 4
Test Number 6 is pair
Test Subscribe 6
Test Number 8 is pair
Test Subscribe 8
Test Number 10 is pair
Test Subscribe 10

Test 2:
Observable.just(1,2,3,4,5,6,7,8,9,10).doOnNext {
            if(it % 2 == 0) {
                Log.v(TAG, "Test 1 Number $it is pair")
            }
        }.filter { it % 2 == 0 }.subscribe {
            Log.v(TAG, "Test 1 Subscribe $it")
        }.addTo(disposables)
Output:
Test 1 Number 2 is pair
Test 1 Subscribe 2
Test 1 Number 4 is pair
Test 1 Subscribe 4
Test 1 Number 6 is pair
Test 1 Subscribe 6
Test 1 Number 8 is pair
Test 1 Subscribe 8
Test 1 Number 10 is pair
Test 1 Subscribe 10

Reply:
@neteinstein you probably meant this
```
doOnNext {
      if (!condition) {
            elseFunction()
      }
 }
 .filter { condition }
```

Reply:
Yep @bruno-garces . Just got confused 😅.
will revert to that.

Reply:
Please validate now @Jmvcaseiro @bruno-garces

Reply:
I commented in a new place